Dockerfile

```diff
@@ -1,33 +1,69 @@
-# see https://pythonspeed.com/articles/alpine-docker-python/ for the reason we don't use alpine
-FROM python:3.9-slim
+# Build stage
+FROM python:3.11-slim as builder
+ENV PATH="/root/.local/bin/:$PATH"
+
 RUN apt-get update \
     && dpkg --add-architecture arm64 \
-    && apt-get install -y --no-install-recommends git ssh socat wget curl libcairo2 python3-dev libffi-dev \
+    && apt-get install -y --no-install-recommends curl gcc \
     && pip3 install --no-cache-dir --upgrade pip \
     && rm -rf /var/lib/apt/lists/*
 
 ENV ENV_TYPE=DEV
 
 RUN mkdir /app
-RUN curl -sSL https://install.python-poetry.org | python3 -
-RUN /root/.local/bin/poetry config virtualenvs.create false
 WORKDIR /app
 
-# Install gcc to compile rumal.yaml.clib, wheel is missing.
-RUN apt-get update \
-    && apt-get install -y --no-install-recommends gcc \
-    && pip3 install --no-cache-dir ruamel.yaml.clib==0.2.8 \
-    && apt-get purge -y --auto-remove gcc \
-    && rm -rf /var/lib/apt/lists/*
+# Create and activate virtual environment
+RUN python -m venv /app/venv --upgrade-deps && \
+    . /app/venv/bin/activate
+
+ENV VIRTUAL_ENV=/app/venv
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 
-# we install the project requirements and install the app in separate stages to optimize docker layer caching
+# Install Poetry
+RUN curl -sSL https://install.python-poetry.org | python3 -
+RUN poetry config virtualenvs.create false
+
+# Install gcc to compile ruamel.yaml.clib, wheel is missing.
+RUN pip3 install --no-cache-dir ruamel.yaml.clib==0.2.8
+
+# Install project dependencies
 COPY pyproject.toml poetry.lock /app/
-RUN /root/.local/bin/poetry install --no-root --no-dev --extras "all"
+RUN poetry install --no-root --without dev --extras "all"
+
+# Copy project source code
 COPY src/ /app/src
-RUN /root/.local/bin/poetry install --no-dev --extras "all"
+RUN poetry install --without dev --extras "all"
```

Comment:
why are we doing `poetry install` twice?

Reply:
this comment was in the original dockerfile
`# we install the project requirements and install the app in separate stages to optimize docker layer caching
`
I think ill add it back in